protocol/encryption/encryptor.go

```diff
@@ -592,6 +596,132 @@ func (s *encryptor) EncryptPayload(theirIdentityKey *ecdsa.PublicKey, myIdentity
+func (s *encryptor) EncryptCommunityPayload(communityId []byte, communityKeyId uint32, payload []byte) (map[string]*EncryptedMessageProtocol, error) {
+	logger := s.logger.With(
+		zap.String("site", "EncryptCommunityPayload"),
+		zap.Any("community-key-id", communityKeyId))
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	logger.Info("encrypting message for communities")
+	dmp, err := s.encryptWithHR(communityId, communityKeyId, payload)
+	response := make(map[string]*EncryptedMessageProtocol)
+	response[noInstallationID] = dmp
+	return response, err
+}
+
 func samePublicKeys(pubKey1, pubKey2 ecdsa.PublicKey) bool {
 	return pubKey1.X.Cmp(pubKey2.X) == 0 && pubKey1.Y.Cmp(pubKey2.Y) == 0
 }
+
+func (s *encryptor) encryptWithHR(groupId []byte, keyId uint32, payload []byte) (*EncryptedMessageProtocol, error) {
+	hrCache, err := s.persistence.GetCommunityKeyByID(groupId, keyId, 0) // Get latest seqNo
+
+	if err != nil {
+		return nil, err
+	}
+
+	s.logger.Info("###encryptWithHR", zap.Any("payload", payload))
+	var dbHash []byte
+	if len(hrCache.Hash) == 0 {
+		s.logger.Info("###encryptWithHR no hash")
+		dbHash = []byte(hrCache.Key)
+	} else {
+		s.logger.Info("###encryptWithHR dbhash", zap.Any("hash", hrCache.Hash))
+		dbHash = hrCache.Hash
+	}
+
+	hash := crypto.Keccak256Hash(dbHash)
+	encryptedPayload, err := crypto.EncryptSymmetric(hash.Bytes(), payload)
+	if err != nil {
+		return nil, err
+	}
+	newSeqNo := hrCache.SeqNo + 1
+	s.logger.Info("###encryptWithHR before SaveCommunityKeyHash", zap.Any("new hash", hash.Bytes()),
+		zap.Any("seqNo", newSeqNo))
+	err = s.persistence.SaveCommunityKeyHash(groupId, keyId, hash.Bytes(), newSeqNo)
+	if err != nil {
+		return nil, err
+	}
+	dmp := &EncryptedMessageProtocol{
+		HRHeader: &HRHeader{
+			GroupId: string(groupId),
+			KeyId:   keyId,
+			SeqNo:   uint32(newSeqNo),
+		},
+		Payload: encryptedPayload,
+	}
+	return dmp, nil
+}
+
+func (s *encryptor) decryptWithHR(groupId []byte, keyId uint32, seqNo uint32, payload []byte) ([]byte, error) {
+
+	hrCache, err := s.persistence.GetCommunityKeyByID(groupId, keyId, seqNo)
+	s.logger.Info("###decryptWithHR", zap.Any("keyId", keyId), zap.Any("hrCache", hrCache), zap.Any("seqNo", seqNo))
+
+	// Handle mesages with seqNo less than the one in db
+	// 1. Check cache. If present for a particular seqNo, all good
+	// 2. Otherwise, get the latest one for that keyId
+	// 3. Every time the key is generated, it has to be saved in the cache along with the hash
+	var hash []byte = hrCache.Hash
+	if hrCache.SeqNo == seqNo {
+		// We already have the hash for this seqNo
+		hash = hrCache.Hash
+	} else {
+		if hrCache.SeqNo == 0 {
+			// No cache records found for this keyId
+			hash = []byte(hrCache.Key)
+		}
+		// We should not have "holes" in seq numbers,
+		// so a case when hrCache.SeqNo > seqNo shouldn't occur
+		for i := hrCache.SeqNo; i < seqNo; i++ {
```

Comment:
I would also leave a not here that we should check that `seqNo` is not too distant from `hrCache.SeqNo`, as that can be used as a vector for an attack.
Say `hrCache.SeqNo` is 1 and a user sends us a message with `seqNo` of MAX_UINT64